csharp/extractor/Semmle.Util/ToolStatusPage.cs

```diff
@@ -0,0 +1,189 @@
+        }
+
+        /// <summary>
+        /// ISO 8601 timestamp.
+        /// </summary>
+        public string Timestamp { get; set; }
+        /// <summary>
+        /// The source of the diagnostic message.
+        /// </summary>
+        public TspSource Source { get; set; }
+        /// <summary>
+        /// GitHub flavored Markdown formatted message. Should include inline links to any help pages.
+        /// </summary>
+        public string? MarkdownMessage { get; set; }
+        /// <summary>
+        /// Plain text message. Used by components where the string processing needed to support
+        /// Markdown is cumbersome.
+        /// </summary>
+        public string? PlaintextMessage { get; set; }
+        /// <summary>
+        /// List of help links intended to supplement <see cref="PlaintextMessage" />.
+        /// </summary>
+        public List<string> HelpLinks { get; }
+        /// <summary>
+        /// SARIF severity.
+        /// </summary>
+        public TspSeverity? Severity { get; set; }
+        /// <summary>
+        /// If true, then this message won't be presented to users.
+        /// </summary>
+        public bool Internal { get; set; }
+        /// <summary>
+        ///
+        /// </summary>
+        public TspVisibility Visibility { get; }
+        public TspLocation Location { get; }
+        /// <summary>
+        /// Structured metadata about the diagnostic message.
+        /// </summary>
+        public Dictionary<string, object> Attributes { get; }
+
+        public DiagnosticMessage(TspSource source)
+        {
+            Timestamp = DateTime.UtcNow.ToString("o", CultureInfo.InvariantCulture);
+            Source = source;
+            HelpLinks = new List<string>();
+            Visibility = new TspVisibility();
+            Location = new TspLocation();
+            Attributes = new Dictionary<string, object>();
+        }
+    }
+
+    /// <summary>
+    /// A wrapper around an underlying <see cref="StreamWriter" /> which allows
+    /// <see cref="DiagnosticMessage" /> objects to be serialized to it.
+    /// </summary>
+    public sealed class DiagnosticsStream : IDisposable
+    {
+        private readonly JsonSerializer serializer;
+        private readonly StreamWriter writer;
+
+        /// <summary>
+        /// Initialises a new <see cref="DiagnosticsStream" /> for a file at <paramref name="path" />.
+        /// </summary>
+        /// <param name="path">The path to the file that should be created.</param>
+        /// <returns>
+        /// A <see cref="DiagnosticsStream" /> object which allows diagnostics to be
+        /// written to a file at <paramname name="path" />.
+        /// </returns>
+        public static DiagnosticsStream ForFile(string path)
+        {
+            var stream = File.CreateText(path);
+            return new DiagnosticsStream(stream);
+        }
+
+        public DiagnosticsStream(StreamWriter streamWriter)
```

Comment:
This leaves an issue with the `IDisposable` contract if the constructor is public as dispose is disposing the `streamWriter`. It seems that the constructor is only used from the `ForFile` static method. Consider putting the stream creation into the constructor.

Reply:
Well spotted, thank you for pointing this out! My original intention was to make this class easier to test, but you're right that this violates the `IDisposable` contract.